Feito por: Olalla Naya Rey

# VISUALIZACIÓN DE MAPAS:

* Este e un mapa interativo:
* Ten marcadas as areas das provincias
* Casas rurais de galicia(extraidas do portal de turismo da xunta de galicia)
* Festas de interese turistico
* E propostas para comer: as nosas propostas.
* Rutas de senderismo: ten rutas sacadas dun artigo, que poñia as rutas mas fermosas da galica maxica e buscadas en wilok, mostrado o trazo. E despois amosamos rutas pgr sacadas do portal de turismo da xunta de galicia

#### Importar librerias

In [ ]:

import geopandas as gpd
import requests
from bs4 import BeautifulSoup
import pandas as pd
import folium
import folium.plugins
import gpxpy
import gpxpy.gpx
import matplotlib.pyplot as plt


#### Ficheiro de provincias de Galicia 



In [ ]:
#https://www.sergas.es/Saude-publica/GIS-provincias'
#paxina donde saquei a carpeta das provincias
# esta o acceso tamen a comarcas por que probei a facelo tamen 
# filepath = 'Comarcas/Comarcas.shp'
# comarcas = gpd.read_file(filepath)

filepath = 'Provincias/Provincias_IGN.shp'
provincias = gpd.read_file(filepath)


#### Casas rurais de galicia:


In [ ]:
#sacado do portal turismo de galicia
#https://www.turismo.gal/localizador-de-recursos/-/sit/onde_aloxarse/turismo_rural/galicia?langId=gl_ES
url='https://www.turismo.gal/localizador-de-recursos?p_p_id=buscador_de_recursos_ritga_portlet_WAR_ritgaportlet&p_p_lifecycle=2&p_p_state=normal&p_p_mode=view&p_p_resource_id=%7B%22orden%22%3A%220%22%2C%22items%22%3A%2210%22%2C%22indexPage%22%3A%221%22%2C%22reverse%22%3A%220%22%2C%22filtroscomunes%22%3A%5B%5D%2C%22filtrosespecificos%22%3A%5B%7B%22metagrupo%22%3A%221%22%2C%22tipologia%22%3A%2271%22%2C%22filtros%22%3A%5B%5D%7D%5D%7D&p_p_cacheability=cacheLevelPage&p_p_col_id=column-1&p_p_col_count=1&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_navegador=KML&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_allFiltro=%26amp%3BpBsq%2523%25231%2523%252371%3Don&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_langId=es_ES&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_langId=es_ES&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_pBsq##1##71=on'

In [ ]:

response = requests.get(url)

soup = BeautifulSoup(response.content,'lxml')

sitio = soup.find_all('placemark')
#sitio

In [ ]:
datos_s=[]
for s in sitio:
    name=s.find('name').text.replace('CDATA','')
    enlace=s.find('a').get('href') #obtemos o enlace
    #cordenadas=f.find('coordinates').text
    lon,lat=s.find('coordinates').text.split(',')[1], s.find('coordinates').text.split(',')[0]
    datos_s.append([name,lon,lat,enlace]) #alamacenamos na lista

datos_st=pd.DataFrame(datos_s,columns=['nome','lon','lat','enlace']) #creamos o dataframe
datos_st
#creamos un dataFrame cos datos necesarios, pero necesitamos formatear o nome sacar todos os caracteres especias.

nome        lon        lat  \
0        <![[ CASA CRUCEIRO]]>   42.09775  -8.334916   
1          <![[A AVOA MARÍA]]>  42.566307  -8.532777   
2               <![[A BOUZA]]>  42.413944  -8.755861   
3              <![[A CAMBOA]]>  42.537167  -8.586861   
4    <![[A CANTARUXA MARUXA]]>   42.20505  -8.458062   
..                         ...        ...        ...   
522           <![[VILAPOMAR]]>  43.485806  -7.066888   
523      <![[VIÑA MEIN S.L.]]>  42.361694  -8.113139   
524         <![[VINOSOBROSO]]>  42.201668  -8.450583   
525       <![[XANELA DA LÚA]]>     42.798  -9.132722   
526           <![[XEITOXIÑA]]>  42.433025  -8.705278   

                                                enlace  
0    http://www.turismo.gal/recurso/-/detalle/2104/...  
1    http://www.turismo.gal/recurso/-/detalle/35940...  
2    http://www.turismo.gal/recurso/-/detalle/15736...  
3    http://www.turismo.gal/recurso/-/detalle/11389...  
4    http://www.turismo.gal/recurso/-/detalle/28038...  
..                                                 ...  
522  http://www.turismo.gal/recurso/-/detalle/32633...  
523  http://www.turismo.gal/recurso/-/detalle/2109/...  
524  http://www.turismo.gal/recurso/-/detalle/36234...  
525  http://www.turismo.gal/recurso/-/detalle/32548...  
526  http://www.turismo.gal/recurso/-/detalle/2175/...  

[527 rows x 4 columns]

In [ ]:
datos_st['nome'] = datos_st['nome'].str.replace('<', '', regex=True).str.replace('[','', regex=True).str.replace(']','', regex=True).str.replace('!','', regex=True).str.replace('>','', regex=True).str.replace('`','', regex=True)
#formateamos correctamente o nome

In [ ]:
datos_st

nome        lon        lat  \
0         CASA CRUCEIRO   42.09775  -8.334916   
1          A AVOA MARÍA  42.566307  -8.532777   
2               A BOUZA  42.413944  -8.755861   
3              A CAMBOA  42.537167  -8.586861   
4    A CANTARUXA MARUXA   42.20505  -8.458062   
..                  ...        ...        ...   
522           VILAPOMAR  43.485806  -7.066888   
523      VIÑA MEIN S.L.  42.361694  -8.113139   
524         VINOSOBROSO  42.201668  -8.450583   
525       XANELA DA LÚA     42.798  -9.132722   
526           XEITOXIÑA  42.433025  -8.705278   

                                                enlace  
0    http://www.turismo.gal/recurso/-/detalle/2104/...  
1    http://www.turismo.gal/recurso/-/detalle/35940...  
2    http://www.turismo.gal/recurso/-/detalle/15736...  
3    http://www.turismo.gal/recurso/-/detalle/11389...  
4    http://www.turismo.gal/recurso/-/detalle/28038...  
..                                                 ...  
522  http://www.turismo.gal/recurso/-/detalle/32633...  
523  http://www.turismo.gal/recurso/-/detalle/2109/...  
524  http://www.turismo.gal/recurso/-/detalle/36234...  
525  http://www.turismo.gal/recurso/-/detalle/32548...  
526  http://www.turismo.gal/recurso/-/detalle/2175/...  

[527 rows x 4 columns]

Temos os datos 526 casas rurais  recomendados polo portal de turismo.

In [ ]:

# o visualizar o mapa por primeira vez comprobamos que unha cordenada faltalle o signo visualizo o dataset e cambioa
df=datos_st[datos_st['nome'] == 'AS ADEGAS']
df
datos_st.iloc[69].replace('8.678661',-8.678661,inplace=True)

#### Paseos que namoran: rutas polas principais fervenzas de galicia.

In [ ]:
#https://www.turismo.gal/localizador-de-recursos?langId=gl_ES&pBsq%23%236%23%2340004370%23%23Alias_tipo_paseo%23%2340000974=on
url5="https://www.turismo.gal/localizador-de-recursos?p_p_id=buscador_de_recursos_ritga_portlet_WAR_ritgaportlet&p_p_lifecycle=2&p_p_state=normal&p_p_mode=view&p_p_resource_id=%7B%22orden%22%3A%220%22%2C%22items%22%3A%2210%22%2C%22indexPage%22%3A%221%22%2C%22reverse%22%3A%220%22%2C%22filtroscomunes%22%3A%5B%5D%2C%22filtrosespecificos%22%3A%5B%7B%22metagrupo%22%3A%226%22%2C%22tipologia%22%3A%2240004370%22%2C%22filtros%22%3A%5B%7B%22alias%22%3A%22Alias_tipo_paseo%22%2C%22valor%22%3A%2240000974%22%7D%5D%7D%5D%7D&p_p_cacheability=cacheLevelPage&p_p_col_id=column-1&p_p_col_count=1&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_navegador=KML&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_allFiltro=%26amp%3BpBsq%2523%25236%2523%252340004370%2523%2523Alias_tipo_paseo%2523%252340000974%3Don"

In [ ]:

response = requests.get(url5)
#print(response.text)
soup = BeautifulSoup(response.content,'lxml')
ruta = soup.find_all('placemark')

In [ ]:
datos=[]
for f in ruta:
    name=f.find('name').text.replace('CDATA','')
    enlace=f.find('a').get('href')
    
    #cordenadas=f.find('coordinates').text
    lon,lat=f.find('coordinates').text.split(',')[1], f.find('coordinates').text.split(',')[0]
    datos.append([name,lon,lat,enlace])

datos_rt=pd.DataFrame(datos,columns=['nome','lon','lat','enlace'])
datos_rt.head()

nome        lon        lat  \
0             <![[FERVENZA DA CIDADELLA]]>  41.885303  -7.216976   
1             <![[FERVENZA DE AUGACAÍDA]]>  42.518154  -7.703239   
2             <![[FERVENZA DE CALLOBRE ]]>  42.692142  -8.429294   
3  <![[FERVENZA DE SANTO ESTEVO DO ERMO]]>   43.53336  -7.238667   
4                <![[FERVENZA DE SEGADE]]>   42.60214  -8.635872   

                                              enlace  
0  http://www.turismo.gal/recurso/-/detalle/21051...  
1  http://www.turismo.gal/recurso/-/detalle/21051...  
2  http://www.turismo.gal/recurso/-/detalle/21051...  
3  http://www.turismo.gal/recurso/-/detalle/21051...  
4  http://www.turismo.gal/recurso/-/detalle/21051...

In [ ]:
datos_rt['nome'] = datos_rt['nome'].str.replace('<', '', regex=True).str.replace('[','', regex=True).str.replace(']','', regex=True).str.replace('!','', regex=True).str.replace('>','', regex=True).str.replace('`','', regex=True)


In [ ]:
datos_rt.shape

(18, 4)

Temos 18 rutas que nos levan a algunhas das fervenzas mais impresionantes de galicia.

#### Festas de interes turistico:

In [ ]:
#url festas
#https://www.turismo.gal/localizador-de-recursos/-/sit/que-facer/festas/galicia?langId=gl_ES
url2="https://www.turismo.gal/localizador-de-recursos?p_p_id=buscador_de_recursos_ritga_portlet_WAR_ritgaportlet&p_p_lifecycle=2&p_p_state=normal&p_p_mode=view&p_p_resource_id=%7B%22orden%22%3A%220%22%2C%22items%22%3A%2210%22%2C%22indexPage%22%3A%221%22%2C%22reverse%22%3A%220%22%2C%22filtroscomunes%22%3A%5B%5D%2C%22filtrosespecificos%22%3A%5B%7B%22metagrupo%22%3A%2226%22%2C%22tipologia%22%3A%2297%22%2C%22filtros%22%3A%5B%5D%7D%5D%7D&p_p_cacheability=cacheLevelPage&p_p_col_id=column-1&p_p_col_count=1&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_navegador=KML&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_allFiltro=%26amp%3BpBsq%2523%252326%2523%252397%3Don&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_pBsq##26##97=on&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_langId=es_ES&_buscador_de_recursos_ritga_portlet_WAR_ritgaportlet_langId=es_ES"

In [ ]:

response = requests.get(url2)
#print(response.text)
soup = BeautifulSoup(response.content,'lxml')
festas = soup.find_all('placemark')


In [ ]:
datos=[]
for f in festas:
    name=f.find('name').text.replace('CDATA','')
    enlace=f.find('a').get('href')
    
    #cordenadas=f.find('coordinates').text
    lon,lat=f.find('coordinates').text.split(',')[1], f.find('coordinates').text.split(',')[0]
    datos.append([name,lon,lat,enlace])

datos_df=pd.DataFrame(datos,columns=['nome','lon','lat','enlace'])
datos_df.head()

nome        lon        lat  \
0  <![[A RAPA DAS BESTAS DE CANDAOSO]]>  43.586945  -7.524999   
1  <![[A RAPA DAS BESTAS DE SABUCEDO]]>  42.593334  -8.414722   
2                     <![[ARDE LUCUS]]>  43.009724     -7.555   
3                   <![[AS SAN LUCAS]]>  43.428333  -7.363333   
4   <![[BELÉN ARTESANAL EN MOVEMENTO]]>  42.705128   -8.67254   

                                              enlace  
0  http://www.turismo.gal/recurso/-/detalle/fi-lu...  
1  http://www.turismo.gal/recurso/-/detalle/fi-po...  
2  http://www.turismo.gal/recurso/-/detalle/fi-lu...  
3  http://www.turismo.gal/recurso/-/detalle/fi-lu...  
4  http://www.turismo.gal/recurso/-/detalle/28038...

In [ ]:
datos_df['nome'] = datos_df['nome'].str.replace('<', '', regex=True).str.replace('[','', regex=True).str.replace(']','', regex=True).str.replace('!','', regex=True).str.replace('>','', regex=True).str.replace('`','', regex=True)


In [ ]:
datos_df.head(2)

nome        lon        lat  \
0  A RAPA DAS BESTAS DE CANDAOSO  43.586945  -7.524999   
1  A RAPA DAS BESTAS DE SABUCEDO  42.593334  -8.414722   

                                              enlace  
0  http://www.turismo.gal/recurso/-/detalle/fi-lu...  
1  http://www.turismo.gal/recurso/-/detalle/fi-po...

#### Restaurantes as nosas propostas

In [ ]:

url= 'https://raw.githubusercontent.com/bigdatawirtz/restaurantes/main/restaurantes.csv'
restaurantes_total = pd.read_csv(url,sep=';')
restaurantes_total.sample(3)

nome   latitude  lonxitude  \
27            O alcume  43.370125  -8.401125   
11    Panadería Caneda  42.339278  -7.253775   
17  Mesón A Canfurnada  43.488688  -8.221579   

                                  comentario  
27            Los calamares están muy buenos  
11  Buen sitio para pillar bicas y empanadas  
17                  Gústanme moito as tostas

#### Ruta senderismo (ca ruta)

In [ ]:
#sacada de https://es.wikiloc.com/
#usamos gpx para leer os archivos gpz
with open('rutas_senderismo/ezaro.gpx', 'r') as gpx_file:
    ezaro = gpxpy.parse(gpx_file)
with open('rutas_senderismo/fontedacoba-cantera-pizarra-a-teixadal.gpx', 'r') as gpx_file:
    fontedacoba = gpxpy.parse(gpx_file)
with open('rutas_senderismo/costa-das-rias-baixas-etapa-13-san-vicente-do-mar-porto-pedr.gpx', 'r') as gpx_file:
    pedras = gpxpy.parse(gpx_file)
with open('rutas_senderismo/subida-al-mustallar-lugo.gpx', 'r') as gpx_file:
    mustallar = gpxpy.parse(gpx_file)
with open('rutas_senderismo/praia-de-area-negra-y-fervenza-en-teixidelo-san-andres-de-te.gpx', 'r') as gpx_file:
    sanandres = gpxpy.parse(gpx_file)
with open('rutas_senderismo/ruta-das-fragas-de-catasos-lalin.gpx', 'r') as gpx_file:
    catasos = gpxpy.parse(gpx_file)
with open('rutas_senderismo/mao-ourense.gpx', 'r') as gpx_file:
    mao = gpxpy.parse(gpx_file)


In [ ]:
#creo unha función para leelos e almacenos nunha lista unha por ruta
def crear(gpx):
    route = []
    for track in gpx.tracks:
        for segment in track.segments:
            for point in segment.points:
                route.append((
                    
                    point.latitude, point.longitude)
                )
    return route

ezaro_list=crear(ezaro)
fondedacoba_list=crear(fontedacoba)
pedras_list=crear(pedras)
mustallar_list=crear(mustallar)
sanandres_list=crear(sanandres)
catasos_list=crear(catasos)
maho_list=crear(mao)

### Mapa

In [ ]:


#m = folium.Map(location=[43, -8.20],zoom_start=8, width=1000, height=1000)
m = folium.Map(location=[43, -8.10],zoom_start=-1,width=1000,height=1000,top=5, left=10) #SE A QUERO GARDAR HTML ASI SE A QUERO VER A LINEA DE ARRIBA
# capa tiles
# https://fontawesome.com/v4/icons/ estas iconas funcionan
#
folium.TileLayer('openstreetmap').add_to(m)
folium.TileLayer('Stamen Toner').add_to(m)
folium.TileLayer('CartoDB Positron').add_to(m)
folium.TileLayer('CartoDB dark_matter').add_to(m)
folium.TileLayer('stamenterrain').add_to(m)
folium.TileLayer('stamen watercolor').add_to(m)


# capa comarcas
folium.GeoJson(provincias, name='Provincias').add_to(m)

# clúster
marker_cluster = folium.plugins.MarkerCluster(name='Festas-Restaurantes').add_to(m)
marker_cluster2=folium.plugins.MarkerCluster(name='Casas Rurales').add_to(m)
marker_cluster3=folium.plugins.MarkerCluster(name='Rutas de Senderismo- Rutas as Fervenzas ', show=False).add_to(m) #asi consigo que non apareza por defecto selecionadada
#capas rutas de senderismo
#con popyline non podemos poñer icona
folium.PolyLine(ezaro_list,color='red', weight=4.5,popup=('Ezaro-MontePindo')).add_to(marker_cluster3)
folium.PolyLine(fondedacoba_list,color='orange', weight=4.5,popup=('Fonte da Coba- Pico dos Formigueiros')).add_to(marker_cluster3)
folium.PolyLine(pedras_list,color='violet', weight=4.5,popup=('Pedras Negras')).add_to(marker_cluster3)
folium.PolyLine(mustallar_list,color='pink', weight=4.5,popup=('Subida o Mustallar')).add_to(marker_cluster3)
folium.PolyLine(sanandres_list,color='yellow', weight=4.5,popup=('Praia da Area Negra - Fervenza do teixidelo')).add_to(marker_cluster3)
folium.PolyLine(catasos_list,color='black', weight=4.5,popup=('Ruta das fragas de Catasos')).add_to(marker_cluster3)
folium.PolyLine(maho_list,color='blue', weight=4.5,popup=('Ruta do Cañon do rio Mao')).add_to(marker_cluster3)
# capa festas
for idx, sitio in datos_df.iterrows():  
    try:
        a_a="<a href=\"" + sitio['enlace']+ "\"target='_blank'>" + sitio['nome']+ "</a> "

        folium.Marker([sitio.lon,sitio.lat],popup= a_a,icon=folium.Icon(color='blue',icon='glyphicon-music')).add_to(marker_cluster)

    except:
        pass
# capa casas rurales
for idx, s in datos_st.iterrows(): 
    try:   
        a="<a href=\"" + s['enlace']+ "\"target='_blank'>" + s['nome']+ "</a> "
        folium.Marker([s.lon,s.lat],popup= a ,icon=folium.Icon(color='green',icon='home')).add_to(marker_cluster2)
    except:
        pass
    
# capa restaurantes as nosas propostas metina noutro cluster, por se solo se quere ir a comer

for idx, restaurante in restaurantes_total.iterrows():
    try:
        folium.Marker([restaurante.latitude,restaurante.lonxitude],
        tooltip=restaurante.nome,
        popup=(f'{restaurante.nome} ({restaurante.comentario})'),
        icon=folium.Icon(color='red',icon='beer', prefix='fa')).add_to(marker_cluster)
    except:
        pass
# fervenzas 
for idx, s in datos_rt.iterrows(): 
    try:   
        a="<a href=\"" + s['enlace']+ "\"target='_blank'>" + s['nome']+ "</a> "
        folium.Marker([s.lon,s.lat],popup= a ,icon=folium.Icon(color='darkgreen',icon='fa-map-signs', prefix='fa')).add_to(marker_cluster3)
    except:
        pass

sw = datos_df[['lon', 'lat']].min().values.tolist()
ne = datos_df[['lon', 'lat']].max().values.tolist()

m.fit_bounds([sw, ne])      
    

folium.LayerControl().add_to(m)
#m

m.save('mapa_rutas.html') #asi gardamolo a html

Ampliación:

* Agregar mais restaurantes posible fonte turismo de galicia.gal ten mais 4000.

* Agregar mais rutas de senderismo.

